File: Java/src/main/java/QuestionBank/Solution901_1000/Solution938.java
package QuestionBank.Solution901_1000;

import tool.TreeNode;

import java.util.ArrayList;

/**
 * @author windows
 */
public class Solution938 {
    ArrayList<Integer> list = new ArrayList<>();
    public int rangeSumBST(TreeNode root, int low, int high) {
        process(root, low, high);
        int sum = 0;
        for (int i = 0; i < list.size(); i++) {
            sum += list.get(i);
        }
        return sum;
    }

    private void process(TreeNode node, int low, int high) {
        if (node == null){
            return;
        }
        if (node.val <= high && node.val >= low){
            list.add(node.val);
        }
        process(node.left, low, high);
        process(node.right, low, high);
    }




    public static void main(String[] args) {
        Solution938 solution938 = new Solution938();
    }
}

File: C++/src/Solution901_1000/Solution938.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution938.h"
#include <functional>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

int rangeSumBST(TreeNode* root, int low, int high) {
  vector<int> nums;
  function<void(TreeNode*)> dfs = [&](TreeNode* node){
    if (node == nullptr){
      return ;
    }
    dfs(node->left);
    nums.push_back(node->val);
    dfs(node->right);
  };
  dfs(root);
  int sum = 0;
  for (int i = 0; i < nums.size(); ++i) {
    if (nums[i] >= low && nums[i] <= high){
      sum += nums[i];
    }
  }
  return sum;
}